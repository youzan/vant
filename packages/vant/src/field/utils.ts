import { HTMLAttributes, InputHTMLAttributes } from 'vue';
import {
  isObject,
  isPromise,
  isFunction,
  getRootScrollTop,
  setRootScrollTop,
} from '../utils';
import type { FieldRule, FieldType, FieldAutosizeConfig } from './types';

export function isEmptyValue(value: unknown) {
  if (Array.isArray(value)) {
    return !value.length;
  }
  if (value === 0 || typeof value === 'boolean') {
    return false;
  }
  return !value;
}

export function runSyncRule(value: unknown, rule: FieldRule) {
  if (isEmptyValue(value)) {
    if (rule.required) {
      return false;
    }
    if (rule.validateEmpty === false) {
      return true;
    }
  }
  if (rule.pattern && !rule.pattern.test(String(value))) {
    return false;
  }
  return true;
}

export function runRuleValidator(value: unknown, rule: FieldRule) {
  return new Promise((resolve) => {
    const returnVal = rule.validator!(value, rule);

    if (isPromise(returnVal)) {
      returnVal.then(resolve);
      return;
    }

    resolve(returnVal);
  });
}

export function getRuleMessage(value: unknown, rule: FieldRule) {
  const { message } = rule;

  if (isFunction(message)) {
    return message(value, rule);
  }
  return message || '';
}

export function startComposing({ target }: Event) {
  target!.composing = true;
}

export function endComposing({ target }: Event) {
  if (target!.composing) {
    target!.composing = false;
    target!.dispatchEvent(new Event('input'));
  }
}

export function resizeTextarea(
  input: HTMLInputElement,
  autosize: true | FieldAutosizeConfig,
) {
  const scrollTop = getRootScrollTop();
  input.style.height = 'auto';

  let height = input.scrollHeight;
  if (isObject(autosize)) {
    const { maxHeight, minHeight } = autosize;
    if (maxHeight !== undefined) {
      height = Math.min(height, maxHeight);
    }
    if (minHeight !== undefined) {
      height = Math.max(height, minHeight);
    }
  }

  if (height) {
    input.style.height = `${height}px`;
    // https://github.com/vant-ui/vant/issues/9178
    setRootScrollTop(scrollTop);
  }
}

export function mapInputType(
  type: FieldType,
  inputmode?: HTMLAttributes['inputmode'],
): {
  type: InputHTMLAttributes['type'];
  inputmode?: HTMLAttributes['inputmode'];
} {
  // type="number" is weird in iOS, and can't prevent dot in Android
  // so use inputmode to set keyboard in modern browsers
  if (type === 'number') {
    type = 'text';
    inputmode ??= 'decimal';
  }

  if (type === 'digit') {
    type = 'tel';
    inputmode ??= 'numeric';
  }

  return { type, inputmode };
}

// get correct length of emoji
// https://github.com/vant-ui/vant/issues/10032
export function getStringLength(str: string) {
  return [...str].length;
}

// cut string with emoji
export function cutString(str: string, maxlength: number) {
  return [...str].slice(0, maxlength).join('');
}
