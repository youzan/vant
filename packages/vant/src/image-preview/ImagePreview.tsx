import {
  ref,
  watch,
  nextTick,
  reactive,
  onMounted,
  defineComponent,
  type PropType,
  type CSSProperties,
  type ExtractPropTypes,
  type TeleportProps,
} from 'vue';

// Utils
import {
  pick,
  truthProp,
  unknownProp,
  Interceptor,
  windowWidth,
  windowHeight,
  makeArrayProp,
  makeStringProp,
  makeNumericProp,
  callInterceptor,
  createNamespace,
  HAPTICS_FEEDBACK,
} from '../utils';

// Composables
import { useRect } from '@vant/use';
import { useExpose } from '../composables/use-expose';

// Components
import { Icon } from '../icon';
import { Swipe, SwipeInstance, SwipeToOptions } from '../swipe';
import { Popup, PopupCloseIconPosition } from '../popup';
import ImagePreviewItem from './ImagePreviewItem';

// Types
import {
  ImagePreviewScaleEventParams,
  ImagePreviewItemInstance,
} from './types';

const [name, bem] = createNamespace('image-preview');

const popupProps = [
  'show',
  'teleport',
  'transition',
  'overlayStyle',
  'closeOnPopstate',
] as const;

export const imagePreviewProps = {
  show: Boolean,
  loop: truthProp,
  images: makeArrayProp<string>(),
  minZoom: makeNumericProp(1 / 3),
  maxZoom: makeNumericProp(3),
  overlay: truthProp,
  vertical: Boolean,
  closeable: Boolean,
  showIndex: truthProp,
  className: unknownProp,
  closeIcon: makeStringProp('clear'),
  transition: String,
  beforeClose: Function as PropType<Interceptor>,
  doubleScale: truthProp,
  overlayClass: unknownProp,
  overlayStyle: Object as PropType<CSSProperties>,
  swipeDuration: makeNumericProp(300),
  startPosition: makeNumericProp(0),
  showIndicators: Boolean,
  closeOnPopstate: truthProp,
  closeOnClickImage: truthProp,
  closeOnClickOverlay: truthProp,
  closeIconPosition: makeStringProp<PopupCloseIconPosition>('top-right'),
  teleport: [String, Object] as PropType<TeleportProps['to']>,
  rotate: Boolean,
};

export type ImagePreviewProps = ExtractPropTypes<typeof imagePreviewProps>;

export default defineComponent({
  name,

  props: imagePreviewProps,

  emits: ['scale', 'close', 'closed', 'change', 'longPress', 'update:show'],

  setup(props, { emit, slots }) {
    const swipeRef = ref<SwipeInstance>();
    const activedPreviewItemRef = ref<ImagePreviewItemInstance>();

    const state = reactive({
      active: 0,
      rootWidth: 0,
      rootHeight: 0,
      disableZoom: false,
      rotateAngles: [] as number[],
    });

    const handleRotate = (direction: 'left' | 'right') => {
      if (!props.rotate) return;
      const angle = 90 * (direction === 'left' ? -1 : 1);
      // 更新当前图片的旋转角度
      state.rotateAngles[state.active] =
        (state.rotateAngles[state.active] || 0) + angle;
    };

    const resize = () => {
      if (swipeRef.value) {
        const rect = useRect(swipeRef.value.$el);
        state.rootWidth = rect.width;
        state.rootHeight = rect.height;
        swipeRef.value.resize();
      }
    };

    const emitScale = (args: ImagePreviewScaleEventParams) =>
      emit('scale', args);

    const updateShow = (show: boolean) => emit('update:show', show);

    const emitClose = () => {
      callInterceptor(props.beforeClose, {
        args: [state.active],
        done: () => updateShow(false),
      });
    };

    const setActive = (active: number) => {
      if (active !== state.active) {
        state.active = active;
        emit('change', active);
      }
    };

    const renderIndex = () => {
      if (props.showIndex) {
        return (
          <div class={bem('index')}>
            {slots.index
              ? slots.index({ index: state.active })
              : `${state.active + 1} / ${props.images.length}`}
          </div>
        );
      }
    };

    const renderCover = () => {
      if (slots.cover) {
        return <div class={bem('cover')}>{slots.cover()}</div>;
      }
    };

    const onDragStart = () => {
      state.disableZoom = true;
    };

    const onDragEnd = () => {
      state.disableZoom = false;
    };

    const renderRotateButtons = () => {
      if (!props.rotate) return null;

      return (
        <div class={bem('rotate-buttons')}>
          <button
            class={bem('rotate-button')}
            onClick={() => handleRotate('left')}
          >
            <Icon
              role="button"
              name="replay"
              class={bem('rotate-icon--reverse')}
            />
          </button>
          <button
            class={bem('rotate-button')}
            onClick={() => handleRotate('right')}
          >
            <Icon role="button" name="replay" />
          </button>
        </div>
      );
    };

    const renderImages = () => (
      <Swipe
        ref={swipeRef}
        lazyRender
        loop={props.loop}
        class={bem('swipe')}
        vertical={props.vertical}
        duration={props.swipeDuration}
        initialSwipe={props.startPosition}
        showIndicators={props.showIndicators}
        indicatorColor="white"
        onChange={setActive}
        onDragEnd={onDragEnd}
        onDragStart={onDragStart}
      >
        {props.images.map((image, index) => (
          <ImagePreviewItem
            v-slots={{
              image: slots.image,
            }}
            rotateAngle={state.rotateAngles[index]}
            ref={(item) => {
              if (index === state.active) {
                activedPreviewItemRef.value = item as ImagePreviewItemInstance;
              }
            }}
            src={image}
            show={props.show}
            active={state.active}
            maxZoom={props.maxZoom}
            minZoom={props.minZoom}
            rootWidth={state.rootWidth}
            rootHeight={state.rootHeight}
            disableZoom={state.disableZoom}
            doubleScale={props.doubleScale}
            closeOnClickImage={props.closeOnClickImage}
            closeOnClickOverlay={props.closeOnClickOverlay}
            vertical={props.vertical}
            onScale={emitScale}
            onClose={emitClose}
            onLongPress={() => emit('longPress', { index })}
          />
        ))}
      </Swipe>
    );

    const renderClose = () => {
      if (props.closeable) {
        return (
          <Icon
            role="button"
            name={props.closeIcon}
            class={[
              bem('close-icon', props.closeIconPosition),
              HAPTICS_FEEDBACK,
            ]}
            onClick={emitClose}
          />
        );
      }
    };

    const onClosed = () => emit('closed');

    const swipeTo = (index: number, options?: SwipeToOptions) =>
      swipeRef.value?.swipeTo(index, options);

    useExpose({
      resetScale: () => {
        activedPreviewItemRef.value?.resetScale();
      },
      swipeTo,
    });

    onMounted(resize);

    watch([windowWidth, windowHeight], resize);

    watch(
      () => props.startPosition,
      (value) => setActive(+value),
    );

    watch(
      () => props.show,
      (value) => {
        const { images, startPosition } = props;
        if (value) {
          state.rotateAngles = images.map(() => 0);
          setActive(+startPosition);
          nextTick(() => {
            resize();
            swipeTo(+startPosition, { immediate: true });
          });
        } else {
          emit('close', {
            index: state.active,
            url: images[state.active],
          });
        }
      },
    );

    return () => (
      <Popup
        class={[bem(), props.className]}
        overlayClass={[bem('overlay'), props.overlayClass]}
        onClosed={onClosed}
        onUpdate:show={updateShow}
        {...pick(props, popupProps)}
      >
        {renderClose()}
        {renderImages()}
        {renderIndex()}
        {renderRotateButtons()}
        {renderCover()}
      </Popup>
    );
  },
});
