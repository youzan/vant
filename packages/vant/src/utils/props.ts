/**
 * prop type helpers
 * help us to write less code and reduce bundle size
 */
import type { PropType } from 'vue';

export const unknownProp = null as unknown as PropType<unknown>;

export const numericProp = [Number, String];

export const truthProp = {
  type: Boolean,
  default: true as const,
};

export const falseProp = {
  type: Boolean,
  default: false as const,
};

export const makeRequiredProp = <T>(type: T) => ({
  type,
  required: true as const,
});

export const makeArrayProp = <T>() => ({
  type: Array as PropType<T[]>,
  default: () => [],
});

export const makeNumberProp = <T>(defaultVal: T) => ({
  type: Number,
  default: defaultVal,
});

export const makeNumericProp = <T>(defaultVal: T) => ({
  type: numericProp,
  default: defaultVal,
});

export const makeStringProp = <T>(defaultVal: T) => ({
  type: String as unknown as PropType<T>,
  default: defaultVal,
});
