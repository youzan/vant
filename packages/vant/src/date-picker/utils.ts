import { extend, padZero, makeArrayProp, clamp } from '../utils';
import { pickerSharedProps } from '../picker/Picker';
import type { PropType } from 'vue';
import type { PickerOption } from '../picker';

type Filter = (
  columnType: string,
  options: PickerOption[],
  values: string[],
) => PickerOption[];
type Formatter = (type: string, option: PickerOption) => PickerOption;

export const sharedProps = extend({}, pickerSharedProps, {
  modelValue: makeArrayProp<string>(),
  filter: Function as PropType<Filter>,
  formatter: {
    type: Function as PropType<Formatter>,
    default: (type: string, option: PickerOption) => option,
  },
});

export const pickerInheritKeys = Object.keys(pickerSharedProps) as Array<
  keyof typeof pickerSharedProps
>;

export function times<T>(n: number, iteratee: (index: number) => T) {
  if (n < 0) {
    return [];
  }

  const result: T[] = Array(n);

  let index = -1;
  while (++index < n) {
    result[index] = iteratee(index);
  }

  return result;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

export const genOptions = <T extends string>(
  min: number,
  max: number,
  type: T,
  formatter: Formatter,
  filter: Filter | undefined,
  values: string[],
) => {
  const options = times(max - min + 1, (index) => {
    const value = padZero(min + index);
    return formatter(type, {
      text: value,
      value,
    });
  });
  return filter ? filter(type, options, values) : options;
};

export const formatValueRange = (
  values: string[],
  columns: PickerOption[][],
) => {
  if (values.length === 0) {
    return columns.map((column) => column[0]?.value?.toString() ?? '');
  }
  return values.map((value, index) => {
    const column = columns[index];
    if (column.length) {
      const minValue = +column[0].value!;
      const maxValue = +column[column.length - 1].value!;
      return padZero(clamp(+value, minValue, maxValue));
    }
    return value;
  });
};
