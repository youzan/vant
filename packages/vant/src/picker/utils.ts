import { isDef, clamp, extend, createNamespace, type Numeric } from '../utils';
import type { Ref } from 'vue';
import type { PickerOption, PickerColumn, PickerFieldNames } from './types';

const [name, bem, t] = createNamespace('picker');

export { name, bem, t };

export const getFirstEnabledOption = (
  options: PickerOption[],
): PickerOption | undefined =>
  options.find((option) => !option.disabled) || options[0];

export function getColumnsType(
  columns: PickerColumn | PickerColumn[],
  fields: Required<PickerFieldNames>,
) {
  const firstColumn = columns[0];
  if (firstColumn) {
    if (Array.isArray(firstColumn)) {
      return 'multiple';
    }
    if (fields.children in firstColumn) {
      return 'cascade';
    }
  }
  return 'default';
}

export function findIndexOfEnabledOption(
  options: PickerOption[],
  index: number,
) {
  index = clamp(index, 0, options.length);

  for (let i = index; i < options.length; i++) {
    if (!options[i].disabled) return i;
  }
  for (let i = index - 1; i >= 0; i--) {
    if (!options[i].disabled) return i;
  }

  return 0;
}

export const isOptionExist = (
  options: PickerOption[],
  value: Numeric | undefined,
  fields: Required<PickerFieldNames>,
) =>
  value !== undefined &&
  options.some((option) => option[fields.value] === value);

export function findOptionByValue(
  options: PickerOption[],
  value: Numeric,
  fields: Required<PickerFieldNames>,
): PickerOption | undefined {
  const index = options.findIndex((option) => option[fields.value] === value);
  const enabledIndex = findIndexOfEnabledOption(options, index);
  return options[enabledIndex];
}

export function formatCascadeColumns(
  columns: PickerColumn | PickerColumn[],
  fields: Required<PickerFieldNames>,
  selectedValues: Ref<Numeric[]>,
) {
  const formatted: PickerColumn[] = [];

  let cursor: PickerOption | undefined = {
    [fields.children]: columns,
  };
  let columnIndex = 0;

  while (cursor && cursor[fields.children]) {
    const options: PickerOption[] = cursor[fields.children];
    const value = selectedValues.value[columnIndex];

    cursor = isDef(value)
      ? findOptionByValue(options, value, fields)
      : undefined;

    if (!cursor && options.length) {
      const firstValue = getFirstEnabledOption(options)![fields.value];
      cursor = findOptionByValue(options, firstValue, fields);
    }

    columnIndex++;
    formatted.push(options);
  }

  return formatted;
}

export function getElementTranslateY(element: Element) {
  const { transform } = window.getComputedStyle(element);
  const translateY = transform.slice(7, transform.length - 1).split(', ')[5];
  return Number(translateY);
}

export function assignDefaultFields(
  fields: PickerFieldNames | undefined,
): Required<PickerFieldNames> {
  return extend(
    {
      text: 'text',
      value: 'value',
      children: 'children',
    },
    fields,
  );
}
