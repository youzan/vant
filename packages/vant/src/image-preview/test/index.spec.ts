import {
  mount,
  later,
  triggerDrag,
  mockGetBoundingClientRect,
  trigger,
} from '../../../test';
import { LONG_PRESS_START_TIME } from '../../utils';
import ImagePreview from '../ImagePreview';
import { images, triggerDoubleTap, triggerZoom } from './shared';
import type { ImagePreviewInstance } from '../types';
import { nextTick } from 'vue';

test('should swipe to current index after calling the swipeTo method', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
  });

  await later();
  (wrapper.vm as Record<string, any>).swipeTo(2);

  await later(100);
  expect(wrapper.find('.van-image-preview__index').html()).toMatchSnapshot();
});

test('should allow to use the teleport prop', () => {
  const root = document.createElement('div');
  mount(ImagePreview, {
    props: {
      show: true,
      teleport: root,
    },
  });

  expect(root.querySelector('.van-image-preview')).toBeTruthy();
});

test('should render cover slot correctly', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
    },
    slots: {
      cover: () => 'Custom Cover',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render index slot correctly', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
    },
    slots: {
      index: ({ index }) => `Custom Index: ${index}`,
    },
  });

  expect(wrapper.find('.van-image-preview__index').html()).toMatchSnapshot();
});

test('should render close icon when using closeable prop', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
      closeable: true,
    },
  });

  wrapper.find('.van-image-preview__close-icon').trigger('click');
  expect(wrapper.emitted<[boolean]>('update:show')![0][0]).toEqual(false);
});

test('should change close icon when using close-icon prop', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      closeable: true,
      closeIcon: 'success',
    },
  });

  expect(
    wrapper.find('.van-image-preview__close-icon').html(),
  ).toMatchSnapshot();
});

test('should change close icon position when using close-icon-position prop', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      closeable: true,
      closeIconPosition: 'top-left',
    },
  });

  expect(
    wrapper.find('.van-image-preview__close-icon').html(),
  ).toMatchSnapshot();
});

test('should hide index when show-index prop is false', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
    },
  });
  expect(wrapper.find('.van-image-preview__index').exists()).toBeTruthy();

  await wrapper.setProps({ showIndex: false });
  expect(wrapper.find('.van-image-preview__index').exists()).toBeFalsy();
});

test('should hide ImagePreview after popstate', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
    },
  });

  window.dispatchEvent(new Event('popstate'));
  expect(wrapper.emitted<[boolean]>('update:show')![0][0]).toBeFalsy();
});

test('should not hide ImagePreview after popstate when close-on-popstate is false', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      closeOnPopstate: false,
    },
  });

  window.dispatchEvent(new Event('popstate'));
  expect(wrapper.emitted('update:show')).toBeFalsy();
});

test('render image', async () => {
  const wrapper = mount(ImagePreview, {
    props: { images, show: true },
  });

  expect(wrapper.html()).toMatchSnapshot();

  await later();

  const swipe = wrapper.find('.van-swipe-item');
  await triggerDrag(swipe, 500, 0);
  expect(wrapper.emitted('update:show')).toBeFalsy();
  expect(wrapper.emitted('change')![0]).toEqual([2]);

  triggerDrag(swipe, 0, 0);
  await later(250);
  expect(wrapper.emitted('update:show')![0]).toEqual([false]);
});

test('before close prop', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      beforeClose: () => true,
    },
  });

  const swipe = wrapper.find('.van-swipe__track');

  // should not emit input or close event when tapped
  triggerDrag(swipe, 0, 0);
  await later(250);
  expect(wrapper.emitted('input')).toBeFalsy();
  expect(wrapper.emitted('close')).toBeFalsy();

  await wrapper.setProps({ show: false });
  expect(wrapper.emitted('close')![0]).toBeTruthy();
});

test('should close when image is clicked', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  await later();
  const image = wrapper.find('.van-image');

  await triggerDrag(image, 0, 0);
  await later(300);
  expect(wrapper.emitted('close')).toBeTruthy();
});

test('should close when overlay is clicked', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  const swipe = wrapper.find('.van-swipe-item');

  await triggerDrag(swipe, 0, 0);
  await later(300);
  expect(wrapper.emitted('close')).toBeTruthy();
});

test('should not close when image is clicked and closeOnClickImage is false', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      closeOnClickImage: false,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  await later();
  const image = wrapper.find('.van-image');

  triggerDrag(image, 0, 0);
  await later(300);
  expect(wrapper.emitted('close')).toBeFalsy();
});

test('should not close when overlay is clicked and closeOnClickOverlay is false', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      closeOnClickOverlay: false,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  const swipe = wrapper.find('.van-swipe-item');

  triggerDrag(swipe, 0, 0);
  await later(300);
  expect(wrapper.emitted('close')).toBeFalsy();
});

test('should trigger scale after double clicking', async () => {
  const onScale = vi.fn();
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      onScale,
    },
  });

  await later();
  const swipe = wrapper.find('.van-swipe-item');

  triggerDoubleTap(swipe);
  expect(onScale).toHaveBeenCalledWith({
    index: 0,
    scale: 2,
  });

  triggerDoubleTap(swipe);
  expect(onScale).toHaveBeenLastCalledWith({
    index: 0,
    scale: 1,
  });

  // when closeOnClickOverlay is set to false, it will not affect the zooming.
  onScale.mockClear();
  await wrapper.setProps({ closeOnClickOverlay: false });
  triggerDoubleTap(swipe);
  expect(onScale).toHaveBeenCalled();
});

test('should allow to disable double click gesture', async () => {
  const onScale = vi.fn();
  const onClose = vi.fn();
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      doubleScale: false,
      onClose,
      onScale,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  // The ImagePreview will close because double-click is disabled.
  await later();
  const swipe = wrapper.find('.van-swipe-item');
  await triggerDoubleTap(swipe);
  expect(onClose).toHaveBeenCalled();
  expect(onScale).not.toHaveBeenCalled();

  // The ImagePreview will not close when closeOnClickOverlay is set to false.
  onClose.mockClear();
  await wrapper.setProps({ closeOnClickOverlay: false, show: true });
  await triggerDrag(swipe, 0, 0);
  expect(onClose).not.toHaveBeenCalled();
  await triggerDoubleTap(swipe, 0, 0);
  expect(onClose).not.toHaveBeenCalled();
});

test('zoom in and drag image to move', async () => {
  const restore = mockGetBoundingClientRect({ width: 100, height: 100 });

  const wrapper = mount(ImagePreview, {
    props: { images, show: true },
  });

  await later();
  const image = wrapper.find('img');
  triggerZoom(image, 300, 300);

  // mock image size
  ['naturalWidth', 'naturalHeight'].forEach((key) => {
    Object.defineProperty(image.element, key, { value: 300 });
  });

  // drag image before load
  triggerDrag(image, 300, 300);
  expect(wrapper.find('.van-image').html()).toMatchSnapshot();

  // drag image after load
  image.trigger('load');
  triggerDrag(image, 300, 300);
  expect(wrapper.find('.van-image').html()).toMatchSnapshot();

  restore();
});

test('zoom out', async () => {
  const restore = mockGetBoundingClientRect({ width: 100, height: 100 });

  const onScale = vi.fn();
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      onScale,
    },
  });

  await later();
  const image = wrapper.find('.van-image');
  triggerZoom(image, 300, 300, 'out');

  expect(onScale).toHaveBeenLastCalledWith({ index: 0, scale: 1 });

  restore();
});

test('should render image slot correctly', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
    slots: {
      image: ({ src }) => `<img class="test-img" src="${src}" />`,
    },
  });

  await later();

  expect(wrapper.html().includes('test-img')).toBeTruthy();
});

test('should render image slot correctly 2', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
    slots: {
      image: ({ src }) =>
        `<video style="width: 100%;" controls><source src="${src}" /></video>`,
    },
  });

  await later();

  expect(wrapper.html().includes('video')).toBeTruthy();
});

test('should render image slot correctly 3', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
    slots: {
      image: ({ src, style }) =>
        `<img class="test-img" src="${src}" style="${Object.assign({ width: '100px' }, style)}" />`,
    },
  });

  await later();

  expect(wrapper.html().includes('width: 100px')).toBeTruthy();
});

test('should emit long-press event after long press', async () => {
  const onLongPress = vi.fn();
  const wrapper = mount(ImagePreview, {
    props: {
      images,
      show: true,
      onLongPress,
    },
  });

  await later();
  const swipe = wrapper.find('.van-swipe-item');
  trigger(swipe, 'touchstart', 0, 0, { x: 0, y: 0 });
  await later(LONG_PRESS_START_TIME + 100);
  trigger(swipe, 'touchend', 0, 0, { touchList: [] });

  expect(onLongPress).toHaveBeenLastCalledWith({
    index: 0,
  });
});

test('should reset scale after calling the resetScale method', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
  });

  await later();
  const image = wrapper.find('.van-image');

  triggerZoom(image, 300, 300);
  await later();
  expect(image.style.transform).toBeTruthy();

  (wrapper.vm as ImagePreviewInstance).resetScale();
  await later();
  expect(image.style.transform).toBeFalsy();
});

test('should render rotate buttons when rotate prop is number', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
      rotate: 90,
    },
  });

  await later();

  const rotateButtons = wrapper.findAll('.van-image-preview__rotate-button');
  expect(rotateButtons).toHaveLength(2);
  expect(
    wrapper.find('.van-image-preview__rotate-buttons').exists(),
  ).toBeTruthy();
});

test('should not render rotate buttons when rotate prop is null', () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
    },
  });

  expect(
    wrapper.find('.van-image-preview__rotate-buttons').exists(),
  ).toBeFalsy();
});

test('should respect custom rotation angle', async () => {
  const wrapper = mount(ImagePreview, {
    props: {
      show: true,
      images,
      rotate: true,
    },
  });
  await later();
  const rotateButtons = wrapper.findAll('.van-image-preview__rotate-button');

  await rotateButtons[1].trigger('click');
  await nextTick();

  const image = wrapper.find('.van-image-preview__image');
  const transformStyle = image.element.style.transform;

  const matches = transformStyle.match(
    /matrix\(([-\d.e+-]+), ([-\d.e+-]+), ([-\d.e+-]+), ([-\d.e+-]+), ([-\d.e+-]+), ([-\d.e+-]+)\)/,
  );
  if (!matches) {
    console.log('No matches found for transform style');
    return;
  }
  const actualMatrix = matches
    .slice(1, 7)
    .map((n: string) => {
      const num = Number(n);
      return Math.abs(num) < 1e-10 ? '0.000000' : num.toFixed(6);
    })
    .join(', ');

  expect(actualMatrix).toBe(
    '0.000000, 1.000000, -1.000000, 0.000000, 0.000000, 0.000000',
  );
});
