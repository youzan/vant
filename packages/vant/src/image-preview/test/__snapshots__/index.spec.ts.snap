// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render image 1`] = `
<transition-stub
  name="van-fade"
  appear="true"
  persisted="false"
  css="true"
>
  <div
    class="van-overlay van-image-preview__overlay"
    role="button"
    tabindex="0"
  >
  </div>
</transition-stub>
<transition-stub
  name="van-fade"
  appear="false"
  persisted="false"
  css="true"
>
  <div
    role="dialog"
    tabindex="0"
    class="van-popup van-popup--center van-image-preview"
  >
    <div class="van-swipe van-image-preview__swipe">
      <div
        style="transition-duration: 300ms; transform: translateX(0px);"
        class="van-swipe__track"
      >
        <div class="van-swipe-item van-image-preview__swipe-item">
        </div>
        <div class="van-swipe-item van-image-preview__swipe-item">
        </div>
        <div class="van-swipe-item van-image-preview__swipe-item">
        </div>
      </div>
    </div>
    <div class="van-image-preview__index">
      1 / 3
    </div>
  </div>
</transition-stub>
`;

exports[`should change close icon position when using close-icon-position prop 1`] = `
"<i class="van-badge__wrapper van-icon van-icon-clear van-image-preview__close-icon van-image-preview__close-icon--top-left van-haptics-feedback" role="button">
  <!---->
  <!---->
  <!---->
</i>"
`;

exports[`should change close icon when using close-icon prop 1`] = `
"<i class="van-badge__wrapper van-icon van-icon-success van-image-preview__close-icon van-image-preview__close-icon--top-right van-haptics-feedback" role="button">
  <!---->
  <!---->
  <!---->
</i>"
`;

exports[`should render cover slot correctly 1`] = `
<transition-stub
  name="van-fade"
  appear="true"
  persisted="false"
  css="true"
>
  <div
    class="van-overlay van-image-preview__overlay"
    role="button"
    tabindex="0"
  >
  </div>
</transition-stub>
<transition-stub
  name="van-fade"
  appear="false"
  persisted="false"
  css="true"
>
  <div
    role="dialog"
    tabindex="0"
    class="van-popup van-popup--center van-image-preview"
  >
    <div class="van-swipe van-image-preview__swipe">
      <div
        style="transition-duration: 300ms; transform: translateX(0px);"
        class="van-swipe__track"
      >
      </div>
    </div>
    <div class="van-image-preview__index">
      1 / 0
    </div>
    <div class="van-image-preview__cover">
      Custom Cover
    </div>
  </div>
</transition-stub>
`;

exports[`should render index slot correctly 1`] = `
<div class="van-image-preview__index">
  Custom Index: 0
</div>
`;

exports[`should swipe to current index after calling the swipeTo method 1`] = `
<div class="van-image-preview__index">
  3 / 3
</div>
`;

exports[`zoom in and drag image to move 1`] = `
<div
  class="van-image van-image-preview__image"
  style="transition-duration: .3s;"
>
  <img
    src="https://fastly.jsdelivr.net/npm/@vant/assets/apple-1.jpeg"
    class="van-image__img"
    style="object-fit: contain;"
  >
  <div class="van-image__loading">
    <div
      class="van-loading van-loading--spinner"
      aria-live="polite"
      aria-busy="true"
    >
      <span class="van-loading__spinner van-loading__spinner--spinner">
        <i class="van-loading__line van-loading__line--1">
        </i>
        <i class="van-loading__line van-loading__line--2">
        </i>
        <i class="van-loading__line van-loading__line--3">
        </i>
        <i class="van-loading__line van-loading__line--4">
        </i>
        <i class="van-loading__line van-loading__line--5">
        </i>
        <i class="van-loading__line van-loading__line--6">
        </i>
        <i class="van-loading__line van-loading__line--7">
        </i>
        <i class="van-loading__line van-loading__line--8">
        </i>
        <i class="van-loading__line van-loading__line--9">
        </i>
        <i class="van-loading__line van-loading__line--10">
        </i>
        <i class="van-loading__line van-loading__line--11">
        </i>
        <i class="van-loading__line van-loading__line--12">
        </i>
      </span>
    </div>
  </div>
</div>
`;

exports[`zoom in and drag image to move 2`] = `
<div
  class="van-image van-image-preview__image"
  style="transition-duration: .3s;"
>
  <img
    src="https://fastly.jsdelivr.net/npm/@vant/assets/apple-1.jpeg"
    class="van-image__img"
    style="object-fit: contain;"
  >
  <div class="van-image__loading">
    <div
      class="van-loading van-loading--spinner"
      aria-live="polite"
      aria-busy="true"
    >
      <span class="van-loading__spinner van-loading__spinner--spinner">
        <i class="van-loading__line van-loading__line--1">
        </i>
        <i class="van-loading__line van-loading__line--2">
        </i>
        <i class="van-loading__line van-loading__line--3">
        </i>
        <i class="van-loading__line van-loading__line--4">
        </i>
        <i class="van-loading__line van-loading__line--5">
        </i>
        <i class="van-loading__line van-loading__line--6">
        </i>
        <i class="van-loading__line van-loading__line--7">
        </i>
        <i class="van-loading__line van-loading__line--8">
        </i>
        <i class="van-loading__line van-loading__line--9">
        </i>
        <i class="van-loading__line van-loading__line--10">
        </i>
        <i class="van-loading__line van-loading__line--11">
        </i>
        <i class="van-loading__line van-loading__line--12">
        </i>
      </span>
    </div>
  </div>
</div>
`;
